storage_service/locations/models/wellcome.py

```diff
@@ -81,29 +85,60 @@ def wellcome_client(self):
     def delete_path(self, delete_path):
         LOGGER.debug('Deleting %s from Wellcome storage', delete_path)
 
-    def move_to_storage_service(self, src_path, dest_path, dest_space):
+    def move_to_storage_service(self, src_path, dest_path, dest_space, package=None):
         """ Moves src_path to dest_space.staging_path/dest_path. """
         LOGGER.debug('Fetching %s on Wellcome storage to %s (space %s)',
             src_path, dest_path, dest_space)
 
-        space_id, source_id = src_path.lstrip('/').split('/')
-
-        bag = self.wellcome_client.get_bag(space_id, source_id)
+        # Possible formats for the path
+        #   /space-id/NAME-uuid.tar.gz
+        #   /space-id/u/u/i/d/NAME-uuid-tar.gz (this happens on reingest)
+        components = src_path.lstrip('/').split('/')
+        space_id, source = components[0], components[-1]
+        filename, ext = source.split('.', 1)
+        name, source_id = filename.split('-', 1)
+
+        # Request a specific bag version if the package has one
+        bag_kwargs = {}
+        if package and 'bag_version' in package.misc_attributes:
+            bag_kwargs['version'] = package.misc_attributes['bag_version']
+
+        # Look up the bag details by UUID
+        bag = self.wellcome_client.get_bag(space_id, source_id, **bag_kwargs)
         loc = bag['location']
         LOGGER.debug("Fetching files from s3://%s/%s", loc['bucket'], loc['path'])
         bucket = self.s3_resource.Bucket(loc['bucket'])
 
-        # The bag is stored unzipped (i.e. as a directory tree).
-        # Download all objects in the source directory
-        s3_prefix = loc['path'].lstrip('/')
+        # The bag contents are stored as individual files on S3, with a
+        # common prefix of the form <space_id>/<bag-id>/<version>
+        # Download all objects with that prefix to a temporary space
+        tmpdir = tempfile.mkdtemp()
+        tmp_aip_dir = os.path.join(tmpdir, filename)
+        s3_prefix = '%s/%s' % (loc['path'].lstrip('/'), bag['version'])
         objects = bucket.objects.filter(Prefix=s3_prefix)
         for objectSummary in objects:
-            dest_file = objectSummary.key.replace(s3_prefix, dest_path, 1)
+            dest_file = os.path.join(
+                tmp_aip_dir,
+                os.path.relpath(objectSummary.key, s3_prefix)
+            )
```

Comment:
This code is quite fragile – the objects in a bag won't necessarily be under `<space>/<bag>/<version>`.

If you do a partial update with a `fetch.txt`, the files might be spread across multiple versions, and this will download an incomplete bag. (And upload it again, too.) I don’t know if Archivematica supports `fetch.txt`; if it does this will cause you issues.

The correct approach is to iterate over the entries in the `manifest` and `tagManifest` in the bag, and download using those keys. That's something that would be useful to have in the Python library, so I'll add that.